tripy/tripy/backend/compiler_api.py

```diff
@@ -31,6 +31,169 @@
 from tripy.common.shape_bounds import ShapeBounds
 from tripy.frontend import Tensor, Trace
 from tripy.utils import json as json_utils
+from tripy.common.device import device as Device
+
+
+# Global variable to store instances
+_default_stream_instances = {}
+
+
+@export.public_api(document_under="compiler")
+def default_stream(device: Device = Device("gpu")) -> "Stream":
+    """
+    Provides access to the default CUDA stream for a given device.
+    This function implements singleton pattern to ensure a single default stream instance per device.
+
+    Args:
+        device (Device, optional): The device for which to get the default stream. Defaults to Device("gpu").
+
+    Returns:
+        Stream: The default Stream object for the specified device.
+
+    Raises:
+        ValueError: If the device is not of type 'gpu' or if the device index is not 0.
+
+    .. code-block:: python
+        :linenos:
+        :caption: Use streams in execution
+
+        default_stream = tp.default_stream()
+        linear = tp.Linear(2, 3)
+        compiler = tp.Compiler(linear)
+
+        compiled_linear = compiler.compile(tp.InputInfo((2, 2), dtype=tp.float32))
+
+        a = tp.ones((2, 2), dtype=tp.float32)
+
+        compiled_linear.stream = default_stream
+        out = compiled_linear(a)
+        default_stream.synchronize()
+        assert cp.array_equal(cp.from_dlpack(out), cp.from_dlpack(linear(a)))
+
+    Usage:
+        default = default_stream()  # Returns the default Stream object for the current device
+        default_gpu1 = default_stream(tp.device("gpu", 1))  # Returns the default Stream object for GPU 1
```

Comment:
I think this should just be the example. We don't necessarily need to show how to assign a stream to an executable here, since that's already covered in `Stream`.

Reply:
Ok, this as an example was a little too short :)